presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -221,38 +228,86 @@ public static HiveSplitSource bucketed(
                             allSplitsLoaded.set(null);
                         }
                     }
 
                     @Override
                     public boolean isFinished(OptionalInt bucketNumber)
                     {
-                        return queueFor(bucketNumber).isFinished();
+                        return splitsFor(bucketNumber).isFinished();
                     }
 
-                    private AsyncQueue<InternalHiveSplit> queueFor(OptionalInt bucketNumber)
+                    @Override
+                    public int rewind(OptionalInt bucketNumber)
+                    {
+                        checkState(allSplitsLoaded.isDone(), "rewind is only compatible when all splits are loaded");
+                        return splitsFor(bucketNumber).rewind();
+                    }
+
+                    private BucketedSplits splitsFor(OptionalInt bucketNumber)
                     {
                         checkArgument(bucketNumber.isPresent());
-                        AtomicBoolean isNew = new AtomicBoolean();
-                        AsyncQueue<InternalHiveSplit> queue = queues.computeIfAbsent(bucketNumber.getAsInt(), ignored -> {
-                            isNew.set(true);
-                            return new AsyncQueue<>(estimatedOutstandingSplitsPerBucket, executor);
-                        });
-                        if (isNew.get() && finished.get()) {
-                            // Check `finished` and invoke `queue.finish` after the `queue` is added to the map.
-                            // Otherwise, `queue.finish` may not be invoked if `finished` is set while the lambda above is being evaluated.
+                        return buckets.computeIfAbsent(bucketNumber.getAsInt(), ignored -> new BucketedSplits(estimatedOutstandingSplitsPerBucket, executor));
+                    }
+
+                    // By adding this synchronization wrapper, we could reduce synchronization granularity to per-bucket level.
+                    class BucketedSplits
+                    {
+                        private final List<InternalHiveSplit> discoveredHiveSplits;
+                        private final AsyncQueue<InternalHiveSplit> queue;
+                        private Optional<ListenableFuture<?>> lastBorrowBatchAsyncFuture = Optional.empty();
+
+                        public BucketedSplits(int estimatedQueueSize, Executor executor)
+                        {
+                            this.discoveredHiveSplits = new ArrayList<>();
+                            this.queue = new AsyncQueue<>(estimatedQueueSize, requireNonNull(executor, "executor is null"));
+                        }
+
+                        public synchronized void offer(InternalHiveSplit internalHiveSplit)
+                        {
+                            discoveredHiveSplits.add(internalHiveSplit);
+                            queue.offer(internalHiveSplit);
```

Comment:
Now I think about it, I think this wrapper class should just return the future, because it is ignored at the caller side anyway and we don't need to make the same comment multiple places.